Add forward AD support for clamp.Tensor

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74205
* __->__ #74042
* #74017

Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

Differential Revision: [D34851567](https://our.internmc.facebook.com/intern/diff/D34851567)

soulitzer committed:
Add forward AD support for clamp when bounds are tensors

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -717,6 +717,57 @@ std::tuple<at::Tensor, at::Tensor> clamp_backward_min_max(
+        self_gte_max.logical_or(max_lt_min) :
+        self_gte_max.logical_or_(max_lt_min);
+
+      return where(pred_self, self_t, scalar_zero) + where(pred_min, min_t, scalar_zero) + where(pred_max, max_t, scalar_zero);
```
As you pointed out in the existing formulas, this is not needed.
You can do `where(self_p < min_p, min_t, where(self_p > max_p, max_t, self_t))` right?

soulitzer replied:
Lol big simplification! I had to nest another where on the outside though to handle the min > max case.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -717,6 +717,57 @@ std::tuple<at::Tensor, at::Tensor> clamp_backward_min_max(
+  } else if (max_p.defined()) {
+    return where(self_p < max_p, self_t, max_t);
+  } else {
+    return self_t; // is this correct?
```
If the function is the identity, yes

soulitzer committed:
Update on "Add forward AD support for clamp.Tensor"


Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for clamp.Tensor"


Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for clamp.Tensor"


Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for clamp.Tensor"


Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -717,6 +717,57 @@ std::tuple<at::Tensor, at::Tensor> clamp_backward_min_max(
+  const Tensor& max_p, const Tensor& max_t
+) {
+  if (min_p.defined() && max_p.defined()) {
+    // We assume self_p is more likely to be tensor-subclass-like than max_p or min_p
```
No longer needed!

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34851567).

soulitzer committed:
Update on "Add forward AD support for clamp.Tensor"


Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

Differential Revision: [D34851567](https://our.internmc.facebook.com/intern/diff/D34851567)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for clamp.Tensor"


Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

Differential Revision: [D34851567](https://our.internmc.facebook.com/intern/diff/D34851567)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for clamp.Tensor"


Trackers: https://github.com/pytorch/functorch/issues/434, https://github.com/pytorch/pytorch/issues/71117
One thing to note about clamp is that when max < min, ignore min and just do clamp_max

Differential Revision: [D34851567](https://our.internmc.facebook.com/intern/diff/D34851567)

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

soulitzer commented:
@pytorchbot merge this
